### ISSUE TITLE
Mysql sum return integer instead of float

### ISSUE THREAD

**Issue by kaluznyo** [rails/rails#12937]
If I do :

```
Variant.sum('public*variants.sold')
```

This request do : 

```
SELECT SUM(public*variants.sold) AS sum_id FROM `variants`
=> 6542
```

And

```
Variant.sum('public*sold')
```

do :

```
SELECT SUM(sold*public) AS sum_id FROM `variants`
=> 6542.34
```

Just to add 'variants.sold' instead of 'sold', the request return a truncate value of float

I use rails 4.0.1, ruby 2, Mysql 5.5 and gem mysql2 (0.3.14)

**Comment by repinel** [comment]
It seems to still be an issue.

```
unless File.exist?('Gemfile')
  File.write('Gemfile', <<-GEMFILE)
    source 'https://rubygems.org'
    gem 'rails', github: 'rails/rails'
    gem 'arel', github: 'rails/arel'
    gem 'sqlite3'
  GEMFILE

  system 'bundle'
end

require 'bundler'
Bundler.setup(:default)

require 'active_record'
require 'minitest/autorun'
require 'logger'

ActiveRecord::Base.establish_connection(adapter: 'sqlite3', database: ':memory:')
ActiveRecord::Base.logger = Logger.new(STDOUT)

ActiveRecord::Schema.define do
  create_table :accounts, force: true  do |t|
    t.integer :value
  end
end

class Account < ActiveRecord::Base
end

class BugTest < Minitest::Test
  def test_association_stuff
    Account.create! value: 2
    Account.create! value: 3
    assert_equal 5.5, Account.sum('1.1 * accounts.value')
  end
end
```
The changes proposed by #13308 would still fix it if they were applied to `type_for` instead of `column_for` since the API changed with #15800

**Comment by robzolkos** [comment]
I tested and this is still an issue in Rails 6.0.3.1     The [PR to fix](https://github.com/rails/rails/pull/13308) was auto-closed due to staleness.

@acapilleri want to take another stab?  Perhaps the use of `sum` in this way doesn't affect enough people.  Would be good to close this issue after so many years.

**Comment by nickborromeo** [comment]
> I tested and this is still an issue in Rails 6.0.3.1

I seems that this is now fixed in `6.1.0.alpha`. I tried the script in https://github.com/rails/rails/issues/12937#issuecomment-101479539 with both sqlite and mysql. 

It seems that this was fixed by https://github.com/rails/rails/pull/39274. @kamipo would you be able to confirm this since you added the fix and close this if that indeed the case? 🙇

**Comment by kamipo** [comment]
I've confirmed this is fixed by #39039, if doesn't find any attribute on the model (and join associations) from `sum(expression)`, aggregated value will be type casted by the aggregated value's type itself.
